### ISSUE TITLE
A pytorch-native ElasticDDP module proposal

### ISSUE THREAD

**Issue by umialpha** [pytorch/elastic#99]
## Motivation/Background

Whenever PET needs to `scale up`, `scale down` or `recover from failure`, the `ElasticAgent` kills and restarts the process.  The major benefit of this approach is that users can leverage elasticity without any change of the code. However, it has some drawbacks.

1. Killing and Restarting typically takes 30 secs or even more. This overhead puts a limit on the frequency of parallelism adjustments. 
2. PET offers the elasticity at the _**epoch granularity**_, which also puts a limit on the frequency of parallelism adjustments. 
## Description

Inspired by the PET V0.1, I would like to suggest adding a pytorch-native `ElasticDDP` module. It has the same distributed training ability as `DDP`, but offers more advanced abilities at the same time.

1. Without `PET` framework, `ElasticDDP` works just like `DDP`.
2. It contains more states that need to be synchronized(e.g. `optimizer`, `current epoch`, `current batch index`).
3. It can (re)initialize the process group,  (re)synchronize states during runtime.
4. It exposes extra APIs to cooperate with PET framework, enabling finer-grained elastic control on the training process, i.e. at `batch` level. 

## Detailed Proposal

![image](https://user-images.githubusercontent.com/13465150/80444185-aa3a8f00-8943-11ea-9c31-4e94cc705c43.png)

Worker

1. **Preparation**. Preparation includes any no-retryable startup logic(model definition, memory allocation, etc.)
2. **Wait Barrier**. After preparation, Worker waits for ElasticAgent's signal.
3. **(Re)Init Process Group**. When the barrier is ready, worker (re)initialize a new process group.
4. **(Re)Sync**. Before training, worker synchronize model states, optimizer states, current batch, current epoch, etc.
5. **Train**. User-provided training function is executed with the state object and any additional user parameters.
6. **Failed?**
 a. If _allreduce_ error occurs, we restore from last committed state and try to rebuild process group.
 b. If other internal error occurs, we exit as failed work.
7. **Commit**. At this point, worker checks whether there is need to re-initialize the process group.
 a. If there is `ScaleUp` signal, go to **(Re)Init Process Group**.
 b. If there is `ScaleDown` signal, go to **(Re)Init Process Group**. We separate this is signal from `ScaleUp` because actually we do not need to **(Re)Sync**.
 c. If there is `Failure` signal, go to **Wait Barrier**.

ElasticAgent
The main logic is similar to the current logic.The differences are:

1. when the group changes( scale up, scale down, failure recovery), it does not kill the worker immediately, instead, it performs the next rendezvous in parallel with Worker training, when it is ready, it sends signal to the Worker, notifying it to reinitialize process group after it commits.
2. When performing the next rendezvous, it must guarantee the oldest Workers occupy the lowest rank, because `ElasticDDP` synchronizes states from rank 0.

### Scale Up
New Node:
Worker does preparation while ElasticAgent performs the next barrier at the same time. When the worker is ready, it waits for the ElasticAgent's signal.

Old Node:
Worker does current training process  while ElasticAgent performs the next barrier at the same time. When Worker `commit`, it checks whether it is time to re-initialize the new process group.
 
### Scale Down
We apply `graceful exit`, When receiving a `SIGTERM` signal. the ElasticAgent will

1. notify other ElasticAgents to performs the next barrier.
2. send `GracefulExit` signal to Worker to stop after current batch.
3. Other ElasticAgents will also send `GracefulExit` signal to their own Worker to stop after current batch.

### Failure

When we `commit`, we can save the  ElasticDDP states into memory(we can control the saving frequencies). When failure occurs, we restore from the last states we saved into memory.

Actually, there are two types of failure, node failure and process failure. 

1. When node fails, the existing ElasticAgent will rebuild the process group and continue the work.
2. When process fails, it is hard to determine which kind of failure has occurred. For similarity, we interpret process failure the same as the node failure.

## Alternatives

1. [Horovod](https://github.com/horovod/horovod/pull/1849) has its first version of elastic under review. I absorb a lot of ideas from its design
2. paper: [Elastic deep learning in multi-tenant GPU cluster](https://arxiv.org/abs/1909.11985) takes a deep insight into elasticity abilities. It is where my initial thought came from.
3. [alibaba-edl](https://github.com/sql-machine-learning/elasticdl) provides k8s-native elasticity for tensorflow.

## TODO
Data access is complicated with elasticity. It needs to discuss in a separate PR.

**Comment by kiukchung** [comment]
Thanks for the proposal! Based on where torchelastic is with v0.2, this proposal  fits better with pytorch distributed more than it does with torchelastic. Feel free to open an RFC on the pytorch github and tag me as a reviewer.
